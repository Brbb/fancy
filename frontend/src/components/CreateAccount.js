import React, { Component } from "react";
import "./CreateAccount.css";
import auth from "../services/auth/api";

class CreateAccount extends Component {
  constructor() {
    super();
    this.state = {
      username: "",
      password: "",
      repeatPassword: "",
      error: ""
    };
  }

  dismissError = () => {
    this.setState({ error: "" });
  };

  handleUserChange = event => {
    this.setState({ username: event.target.value });
  };
  handlePassChange = event => {
    this.setState({ password: event.target.value });
  };
  handleRepeatPassChange = event => {
    if (this.state.password !== event.target.value) {
      this.setState({ error: "Passwords must match" });
    } else this.dismissError();
    this.setState({ repeatPassword: event.target.value });
  };

  signup = async () => {
    var signupResult = await auth.signup(
      this.state.username,
      this.state.password,
      this.state.repeatPassword
    );
    if (signupResult.success) {
      this.props.history.push("/");
    } else this.setState({ error: signupResult.reason });
  };

  render() {
    return (
      <div className="CreateAccount formContainer">
        <form className="form" onSubmit={this.handleSubmit}>
          <label className="formTitle">Fancy App</label>
          <label>Email</label>
          <input
            type="text"
            className="textField"
            placeholder="user@email.com"
            data-test="username"
            value={this.state.username}
            onChange={this.handleUserChange}
          />

          <label>Password</label>
          <input
            type="password"
            placeholder="super secret"
            data-test="password"
            className="textField"
            value={this.state.password}
            onChange={this.handlePassChange}
          />
          <input
            type="password"
            placeholder="repeat super secret"
            data-test="password"
            className="textField"
            value={this.state.repeatPassword}
            onChange={this.handleRepeatPassChange}
          />
          {this.state.error && (
            <label
              className="loginError"
              data-test="error"
              onClick={this.dismissError}
            >
              {this.state.error}
            </label>
          )}
          <input
            className="submitBtn"
            type="button"
            value="Create New Account"
            onClick={this.signup}
            disabled={
              this.state.username === "" ||
              this.state.password === "" ||
              this.state.repeatPassword === "" ||
              this.state.error !== ""
            }
          />
        </form>
      </div>
    );
  }
}

export default CreateAccount;
