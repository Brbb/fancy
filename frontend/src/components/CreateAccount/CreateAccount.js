import React, { Component } from "react";
import "./CreateAccount.css";
import auth from "../../services/auth/api";
import { Button, InputField, Message, SectionTitle } from "../Elements/Basics";
import PasswordGroup from "../Elements/PasswordGroup";

class CreateAccount extends Component {
  constructor() {
    super();
    this.state = {
      username: "",
      password: "",
      repeatPassword: "",
      error: ""
    };
  }

  handleUserChange = event => {
    this.setState({ username: event.target.value });
  };

  signup = async () => {
    var result = await auth.signup(
      this.state.username,
      this.state.password,
      this.state.repeatPassword
    );
    if (!result.err) {
      this.props.history.push("/");
    } else this.setState({ error: result.err });
  };

  render() {
    return (
      <div className="f-section bordered centered">
        <SectionTitle text="New Account" />
        <InputField
          name="emailField"
          placeholder="user@email.com"
          value={this.state.username}
          onChange={this.handleUserChange}
          text="Email"
        />
        <PasswordGroup
          onPasswordsChange={(newPassword, repeatPassword, error) => {
            this.setState({
              password: newPassword,
              repeatPassword: repeatPassword,
              error: error
            });
          }}
        />
        <Button
          name="create-button"
          className="f-submit-button"
          text="Create New Account"
          onClick={this.signup}
          disabled={
            this.state.username === "" ||
            this.state.password === "" ||
            this.state.repeatPassword === "" ||
            this.state.error !== ""
          }
        />
        <Button
          className="f-button"
          text="← Back"
          onClick={() => this.props.history.push("/")}
        />
        <Message text={this.state.error} success={this.state.error === ""} />
      </div>
    );
  }
}

export default CreateAccount;
