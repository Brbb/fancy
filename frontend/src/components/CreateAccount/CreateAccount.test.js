import CreateAccount from "./CreateAccount";
import authApi from "../../services/auth/api";

describe("CreateAccount component", () => {
  describe("Rendering tests", () => {
    test("CreateAccount mounts and fields are initialized correctly", () => {
      let wrapper = shallow(<CreateAccount />);
      let createButton = wrapper.find('[name="create-button"]').first();

      expect(wrapper).toMatchSnapshot();
      expect(wrapper.state().password).toEqual("");
      expect(wrapper.state().repeatPassword).toEqual("");
      expect(wrapper.state().error).toEqual("");
      expect(createButton.prop("disabled")).toBeTruthy();
    });

    test("Save button gets enabled correctly on field/state changes", () => {
      let wrapper = mount(<CreateAccount />);
      wrapper
        .find('[name="emailField"]')
        .last()
        .simulate("change", { target: { value: "test@email.com" } });
      wrapper
        .find('[name="newPasswordField"]')
        .last()
        .simulate("change", { target: { value: "pass123" } });
      wrapper
        .find('[name="repeatNewPasswordField"]')
        .last()
        .simulate("change", { target: { value: "pass123" } });

      expect(wrapper).toMatchSnapshot();
      expect(wrapper.state().username).toEqual("test@email.com");
      expect(wrapper.state().password).toEqual("pass123");
      expect(wrapper.state().repeatPassword).toEqual("pass123");

      // re-find the button to include re-rendered prop
      let saveButton = wrapper.find('[name="create-button"]').first();
      expect(saveButton.prop("disabled")).toBeFalsy();

      wrapper
        .find('[name="repeatNewPasswordField"]')
        .last()
        .simulate("change", { target: { value: "wrong" } });
      expect(wrapper.state().repeatPassword).toEqual("wrong");
      expect(wrapper.state().error).toEqual("Passwords must match");
    });
  });

  describe("Behavioral test", () => {
    test("Create button calls signup and on success goes to login", () => {
      let location = { pathname: "/new" };
      let historyMock = {
        push: jest.fn(path => {
          location.pathname = path;
        })
      };

      let wrapper = shallow(<CreateAccount history={historyMock} />);
      jest.spyOn(authApi, "signup").mockImplementation(() => {
        return Promise.resolve({});
      });

      expect(authApi.signup).toBeCalledTimes(0);
      expect(location.pathname).toBe("/new");

      let createButton = wrapper.find('[name="create-button"]').first();
      return createButton
        .prop("onClick")()
        .then(() => {
          expect(authApi.signup).toBeCalledTimes(1);
          expect(location.pathname).toBe("/");
          jest.clearAllMocks();
        });
    });

    test("Save button calls updatePassword and on error shows a message", () => {
      let location = { pathname: "/new" };
      let historyMock = {
        push: jest.fn(path => {
          location.pathname = path;
        })
      };

      let wrapper = shallow(<CreateAccount history={historyMock} />);
      jest.spyOn(authApi, "signup").mockImplementation(() => {
        return Promise.resolve({ err: "Any" });
      });

      expect(authApi.signup).toBeCalledTimes(0);
      expect(location.pathname).toBe("/new");

      let createButton = wrapper.find('[name="create-button"]').first();
      return createButton
        .prop("onClick")()
        .then(() => {
          let messageField = wrapper.find("Message").first();
          expect(authApi.signup).toBeCalledTimes(1);
          expect(location.pathname).toBe("/new");
          expect(wrapper.state().error).toBe("Any");
          expect(messageField.prop("text")).toBe("Any");
          jest.clearAllMocks();
        });
    });
  });
});
