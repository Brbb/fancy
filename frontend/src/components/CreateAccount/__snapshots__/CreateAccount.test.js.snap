// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccount component Rendering tests CreateAccount mounts and fields are initialized correctly 1`] = `
<div
  className="f-section bordered centered"
>
  <SectionTitle
    text="New Account"
  />
  <InputField
    name="emailField"
    onChange={[Function]}
    placeholder="user@email.com"
    text="Email"
    value=""
  />
  <PasswordGroup
    onPasswordsChange={[Function]}
  />
  <Button
    className="f-submit-button"
    disabled={true}
    name="create-button"
    onClick={[Function]}
    text="Create New Account"
  />
  <Button
    className="f-button"
    onClick={[Function]}
    text="← Back"
  />
  <Message
    success={true}
    text=""
  />
</div>
`;

exports[`CreateAccount component Rendering tests Save button gets enabled correctly on field/state changes 1`] = `
<div
  className="f-section bordered centered"
>
  <div
    className="f-section-title"
  >
    <label>
      New Account
    </label>
  </div>
  <div
    className="f-input"
  >
    <label
      className="f-input-field-label"
    >
      Email
    </label>
    <input
      className="f-input-field"
      name="emailField"
      onChange={[Function]}
      placeholder="user@email.com"
      value="test@email.com"
    />
  </div>
  <div>
    <div
      className="f-input"
    >
      <label
        className="f-input-field-label"
      >
        Password
      </label>
      <input
        className="f-input-field"
        name="newPasswordField"
        onChange={[Function]}
        placeholder="Super Secret"
        type="password"
        value="pass123"
      />
    </div>
    <div
      className="f-input"
    >
      <label
        className="f-input-field-label"
      />
      <input
        className="f-input-field"
        name="repeatNewPasswordField"
        onChange={[Function]}
        placeholder="Repeat Super Secret"
        type="password"
        value="pass123"
      />
    </div>
  </div>
  <input
    className="f-submit-button"
    disabled={false}
    name="create-button"
    onClick={[Function]}
    type="button"
    value="Create New Account"
  />
  <input
    className="f-button"
    onClick={[Function]}
    type="button"
    value="← Back"
  />
  <label
    className="f-message f-message-s"
  />
</div>
`;
